src/query/service/src/pipelines/processors/transforms/hash_join/result_blocks.rs

```diff
@@ -60,115 +62,171 @@ impl JoinHashTable {
         Key: HashTableKeyable + Clone + 'static,
         IT: Iterator<Item = Key> + TrustedLen,
     {
-        let probe_indexs = &mut probe_state.probe_indexs;
-        let build_indexs = &mut probe_state.build_indexs;
         let valids = &probe_state.valids;
 
-        let mut results: Vec<DataBlock> = vec![];
         match self.hash_join_desc.join_type {
             JoinType::Inner => {
+                let block_size = self
+                    .ctx
+                    .get_settings()
+                    .get_max_block_size()
+                    .unwrap_or(65535) as usize;
+
+                // The inner join will return multiple data blocks of similar size
+                let mut probed_blocks = vec![];
+                let mut probe_indexes = Vec::with_capacity(block_size);
+                let mut build_indexes = Vec::with_capacity(block_size);
+
                 for (i, key) in keys_iter.enumerate() {
                     // If the join is derived from correlated subquery, then null equality is safe.
                     let probe_result_ptr = if self.hash_join_desc.from_correlated_subquery {
                         hash_table.find_key(&key)
                     } else {
                         Self::probe_key(hash_table, key, valids, i)
                     };
-                    match probe_result_ptr {
-                        Some(v) => {
-                            let probe_result_ptrs = v.get_value();
-                            build_indexs.extend_from_slice(probe_result_ptrs);
 
-                            for _ in probe_result_ptrs {
-                                probe_indexs.push(i as u32);
+                    if let Some(v) = probe_result_ptr {
```

Comment:
Help me check if I understand the logic right @zhang2014 

By splitting the probe phase into multi-parts, check if `interrupt` is true during each part to make probe can break as soon as possible if query was killed.

If my understanding is correct, maybe we also need the similar logic to other join's probe phase, such as mark join, semi join.  The case in the issue https://github.com/datafuselabs/databend/issues/8028 is a semi join.

Reply:
> Help me check if I understand the following logic right @zhang2014
> 
> By splitting the probe phase into multi-parts, check if is true during each part to make probe can break as soon as possible if query was killed.`interrupt`

You're right. 

> If my understanding is correct, maybe we also need the similar logic to other join's probe phase, such as mark join, semi join?
> 
> The case in the issue #8028 is a semi join.

This will be implemented in the next pr.